api.py

```diff
@@ -79,6 +83,31 @@ def post(self, url, data=None, json=None, **kwargs):
 
         return super().post(url, data=data, json=json, **kwargs)
 
+    def delete(self, url):
+        super().delete(url)
+
+        # Wait for snapshots to be deleted
+        if 'volume-snapshots' in url:
+            timeout = datetime.now() + timedelta(seconds=60)
+
+            while datetime.now() < timeout:
+                time.sleep(1)
+
+                try:
+                    snapshot = self.get(url)
+                except requests.exceptions.HTTPError as e:
+                    if e.response.status_code == 404:
+                        # The snapshot is gone, stop waiting
+                        break
+                    else:
+                        raise e
+
+            else:
+                raise Timeout(
```

Comment:
I think if we are going to raise a timeout in a delete function, the cleanup should catch that and try to delete as much else as possible. Now, a single snapshot running into a timeout ensures that none of the other resources are cleaned up, iiuc.

Technically all the other resources can have similar issues (with exceptions), so this is not necessary a problem introduced with this change.

Reply:
I improved the cleanup code in general to collect all exceptions and  re-raise them later.